File: cpp/trapn_cc.cpp

#include <iostream>
#include <iomanip>
#include <cmath>
#include <omp.h>

const double pi = M_PI;

double transform_cc(double t)
{
    double L = 1.0;

    // Can cause performance problems (underflow?) near pi/2
    //return L/std::tan(t);

    // Use alternate formulation for cotangent
    double sin_t;
    double cos_t;
    sincos(t, &sin_t, &cos_t);
    return L*cos_t/sin_t;
}

double jacobian_cc(double t)
{
    double L = 1.0;
    double s = std::sin(t);
    return L/(s*s);
}

const int ndim = 6;

double fn(double xx[ndim])
{
    double x = xx[0];
    double y = xx[1];
    double z = xx[2];
    double x2 = xx[3];
    double y2 = xx[4];
    double z2 = xx[5];
    return exp(-x*x) * exp(-2*y*y) * exp(-z*z) * exp(-x2*x2)* exp(-y2*y2) * exp(-z2*z2);
}

// https://stackoverflow.com/questions/46782444/how-to-convert-a-linear-index-to-subscripts-with-support-for-negative-strides
void ind2sub(int idx, const int* shape, int* indices)
{
    for (int i = 0; i < ndim; i++)
    {
        int s = idx % shape[i];
        idx -= s;
        idx /= shape[i];
        indices[i] = s;
    }
}


template <typename F>
double trapn_cc(int n, F &f)
{
    double h = pi/(n+1);

    double total = 0.0;

    int npts = std::pow(n, ndim);
    int indices[ndim];
    int nn1[ndim];
    for (int i = 0; i < ndim; i++) {
        nn1[i] = n;
    }

#pragma omp parallel for reduction(+:total)
    for (int i = 0; i < npts; i++) {
        double xx[ndim];
        ind2sub(i, nn1, indices);
        double jac = 1.0;
        for (int j = 0; j < ndim; j++) {
            // Map to (0,pi)
            double x = (indices[j]+1) * h;

            // Map to (-oo, oo)
            xx[j] = transform_cc(x);

            jac *= jacobian_cc(x);
        }
        total += jac*f(xx);
    }

    return total*std::pow(h, ndim);
}

int main()
{
    //int n = 16;
    double expected = std::sqrt(pi/2) * std::pow(std::sqrt(pi), 5);
    std::cout << "# expected = " << expected << std::endl;
    std::cout << "# n      val     diff    time" << std::endl;
    for (int n = 10; n <= 24; n++) {
        double start = omp_get_wtime();
        double val = trapn_cc(n, fn);
        double end = omp_get_wtime();
        double diff = expected-val;
        std::cout << n << " " << val << " " << diff << " " << end-start << std::endl;
    }

}

File: python/trapn_cc.py

from math import exp, pi, tan, sin, sqrt
import numpy as np
# Clenshaw-Curtis - trapezoidal rule from -oo to oo with n divisions
# https://en.wikipedia.org/wiki/Clenshaw%E2%80%93Curtis_quadrature

# In N dimensions

# Coordinate transformation
@np.vectorize
def transform_cc(t):
    # problem-dependent convergence factor
    L = 1.0
    # There may be performance or numerical issues around pi/2.  Use
    # a different formula for cotangent in that case.
    return L/tan(t)

# Jacobian of the transformation
@np.vectorize
def jacobian_cc(t):
    # problem-dependent convergence factor
    L = 1.0
    return L/sin(t)**2

# https://stackoverflow.com/questions/46782444/how-to-convert-a-linear-index-to-subscripts-with-support-for-negative-strides
def ind2sub(idx, shape, indices):
    for i in range(len(shape)):
        s = idx % shape[i]
        idx -= s
        idx //= shape[i]
        indices[i] = s
    return indices

# Assume the integrand goes to zero faster than 1/x^2, so the endpoints evaluate to zero.

# n is the number of interior grid points.
# The number of intervals is n+1
# Including the end points, the number of grid points is n+2
#   Since they contribute zero to the integral, we can skip them and
#   only use the interior grid points.

def trapn_cc(ndim, n, fn):
    a = 0.0
    b = pi

    # Size of each interval
    h = (1.0/(n+1))*(pi)
    hval = h**ndim

    total = 0.0
    nn1 = [n]*ndim
    idx = np.zeros(ndim, dtype=np.int64)
    x = np.zeros(ndim)
    xt = np.zeros(ndim)

    npts = n**ndim
    for i in range(npts):
        idx = ind2sub(i, nn1, idx) + 1

        # Map to (0,pi)
        x[:] = idx*h

        # Map to (-oo,oo)
        xt[:] = transform_cc(x)
        jac = np.prod(jacobian_cc(x))
        total += jac*fn(xt)

    return total*hval

# Integral is -exp(-x)
def exp_fn(x):
    x1 = x[0]
    x2 = x[1]
    x3 = x[2]
    x4 = x[3]
    x5 = x[4]
    x6 = x[5]
    return exp(-x1*x1)*exp(-2*x2*x2)*exp(-3*x3*x3)*exp(-x4*x4)*exp(-x5*x5)*exp(-x6*x6)


if __name__ == '__main__':

    # Number of grid points used in each dimension
    n = 6

    e = sqrt(pi) * sqrt(pi/2) * sqrt(pi/3) * sqrt(pi) * sqrt(pi) * sqrt(pi)
    print('Expected value: ',e)
    # ---------------------
    # Exponential integrand
    # ---------------------
    val = trapn_cc(6, n, exp_fn)
    print('Integrated value: ',val, ' Difference: ',val-e)

